modules/gapi/src/executor/thread_pool.cpp

```diff
@@ -0,0 +1,88 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+//
+// Copyright (C) 2024 Intel Corporation
+
+
+#include "thread_pool.hpp"
+
+#include <opencv2/gapi/util/throw.hpp>
+
+void cv::gapi::own::WaitGroup::add() {
+    std::lock_guard<std::mutex> lk{m};
```

Comment:
In fact it's not necessarily to acquire the mutex on `add()` method, `task_counter` might be an `atomic`.

But since we still need the `mutex` for `wait` (as it's uses condvar) `atomic` + `mutex` seems the weird combination and didn't give any significant performance improvements. Decided to keep it simple so far.

If fact `WaitGroup` can be implemented without condvar and mutex by using 2 atomics. But it requires c++20 for this thing: https://en.cppreference.com/w/cpp/atomic/atomic/wait